### ISSUE TITLE
GPS time stamp 

### ISSUE THREAD

**Issue by caraMsmith** [zross/sensorDataImport#36]
Hi Zev, the time stamps for all the gps files are off by 4 hours. When I view the gpx files in google earth the correct times appears, but in the database they are 4 hours later. The time stamps for the other files types are correct, only the gps (.gpx) files are different. When the gpx files are opened in a text edit application, the times are 4 hours later. My guess is google earth is able to recognize and adjust to the correct timezone, but the database is reading as is directly from the file text. Is there something we can do to adjust the times? 

Thanks!

**Comment by dwjack** [comment]
I think that this is because GPX uses UTC time - see http://www.topografix.com/gpx/1/1/gpx.xsd, which says "Date and time in are in Univeral Coordinated Time (UTC), not local time! Conforms to ISO 8601 specification for date/time representation."

Seems that we need to do one of three things:

1)  have user enter time zone at upload;
2)  parse the first lat/lon for the correct time zone (e.g., http://stackoverflow.com/questions/16086962/how-to-get-a-time-zone-from-a-location-using-latitude-and-longitude-coordinates)
3)  Map projects to time zones -- assume that all the biking data is from NYC (but even then we will need to handle daylight savings).

What are your thoughts Zev?

**Comment by zross** [comment]
Shucks, I had assumed that the time zone for all files was the same. As a result, I deliberately set the system NOT to use a time zone, #8, because there are challenges to using timezones related to the fact that the server has one timezone, the user has another and R could have another. Since I thought an entire project could be self-contained and all the files have the same time zone then there would be no need for TZ -- I guess not...

Here is the proposed fix:

1) Set the datetime field in all tables to be `TIMESTAMP WITH TIME ZONE NOT NULL`
2) Add a function that returns a timezone based on a project name -- this will be hard-coded
3) Alter the `upload_postgres()` function to append the project timezone to all files EXCEPT GPS. For GPS assign the UTC time zone.
4) [I think this is required] Alter the `get_sensor_data()` function to ensure that the displayed date is in the PROJECT time zone not the local time zone.

@dwjack, since you're on the W Coast, can you test something for me to tell me if item 4 above is necessary?

**Create a table in PostgreSQL**

```sql
create table timezonetest(
  datetime TIMESTAMP WITH TIME ZONE NOT NULL
);
```

**INSERT two rows of data**

```sql
INSERT INTO timezonetest (datetime) VALUES ('2015-06-19 12:21:30 UTC');
INSERT INTO timezonetest (datetime) VALUES ('2015-06-19 08:21:30 America/New_York');
```

**Take a look at the results in the table**

I'm guessing this will be something like:

2015-06-19 05:21:30-07


** Pull the data into R with**

You have the sensorDataImport package installed?

```r
get_connection("columbiaBike","spatial",  host="localhost",
               port=5433, user="postgres")
x<-collect(tbl(.connection, "timezonetest"))
format(x[,1], usetz = TRUE)
```

** Let me know what the `format()` function above shows**

Perhaps paste the results for me?

**Comment by zross** [comment]
@caraMsmith Thanks for paying attention and reporting this. If it would take a significant amount of time to re-upload I can give you a few lines of SQL and tell you how to run it. For the GPS table we would use:

```sql
ALTER TABLE gps ALTER datetime TYPE timestamptz USING (datetime AT TIME ZONE 'UTC')
```

And then for the other tables we would use:

```sql
ALTER TABLE hxi ALTER datetime TYPE timestamptz USING (datetime AT TIME ZONE 'America/New_York')
```

**Comment by caraMsmith** [comment]
@zross Do I need to do the steps in your first comment first, or just the steps in the second comment.

**Comment by zross** [comment]
I'll need to update the package eventually but short-term you can do the following steps. Keep in mind that when R reads in the data the data display will show the database times in local time (if it's 6pm EST in the database, it will show up as 3pm PST if you're in CA). This is fine if you're working in New York on the New York project, but I'll need to get the package set up for, say, the Ghana project. If this is a problem short-term let's discuss by phone.

1) backup the database: I'm assuming Darby showed you pgAdmin III? Open this program, single click on the columbiaBike database. Then right click and choose backup, choose a name and path, otherwise accept the defaults.

2) click on the "SQL" button with the magnifying glass. Run:

```sql
ALTER TABLE gps ALTER datetime TYPE timestamptz USING (datetime AT TIME ZONE 'UTC');
ALTER TABLE abp ALTER datetime TYPE timestamptz USING (datetime AT TIME ZONE 'America/New_York');
ALTER TABLE hxi ALTER datetime TYPE timestamptz USING (datetime AT TIME ZONE 'America/New_York');
ALTER TABLE mae ALTER datetime TYPE timestamptz USING (datetime AT TIME ZONE 'America/New_York');
ALTER TABLE mpm ALTER datetime TYPE timestamptz USING (datetime AT TIME ZONE 'America/New_York');
```

**Comment by caraMsmith** [comment]
I backed up the database on pgAdmin lll and ran the SQL code. I then deleted and re-loaded one gpx file. The time is still off for that file. Since there were changes in pgAdmin lll do I need to reinstall the database? Or is there something else I need to do?

**Comment by zross** [comment]
Sorry, I thought you were done with uploading and this change was for analysis only. I have made a change to the upload function to append the timezone. You can re-install the package with 

`devtools::install_github("zross/sensorDataImport")`

And the upload should work. I will still need to make changes to the processing scripts for people working in a different timezone but should be fine for you for now.

FYI -- I also made a change to the data extraction function so that you can extract raw data (there was a bug before)

**Comment by caraMsmith** [comment]
It worked! Thanks Zev!

**Comment by dwjack** [comment]
Thanks Zev and Cara.

First of all, I tried to follow your instructions Zev for 'timezonetest'.  After fiddling for a bit I couldn't get it to work (errors in R, perhaps related to running it on a mac?)

Just to make sure that I understand where we came to rest on this:

1)  The updated sensorDataImport package now assigns a TZ based on the project, so future imports should be correct;
2)  The SQL commands that you gave above should fix the existing GPX data.  

@caraMsmith, were you able to verify that both 1 & 2 are working?

**Comment by zross** [comment]
Yes, you should be right on 1 and 2 (though Cara can confirm that these work). So we're all set with a short-term fix for the NYC project.

But in a comment above I listed four steps that would need to be done to complete the loop and I have not done 2 or 4. 

So right now the upload function is appending the timezones associated with the NYC bike project and will not do the right thing with a project, like Ghana, in another timezone. And, if you, in CA, were to extract the data it might read as PST rather than EST. I can explain further. These are not very time consuming but would take a little time to straighten out. I am going to create a new "issue" listing my priorities for the remaining budget. This is probably one of them.

**Comment by caraMsmith** [comment]
1) Yes
2) Kind of. I had to re-upload the previously uploaded gpx files, so now they all have the correct time. Was it supposed to fix the files without me re-uploading them? After I ran the SQL code nothing changed until I reinstalled the package, deleted the gpx files, and then reloaded them. 
Either way, they have the correct times now.

**Comment by zross** [comment]
Did you "refresh" after running the SQL code? But after running you need to refresh your data view to see the changes. The SQL code was supposed to make it so you did not have to re-upload and worked for me. It's also a little confusing because PostgreSQL displays dates/times in the your current timezone and then adds a little "-04" to show you that it's being stored in UTC.

In any case, glad times are correct now, sorry that you re-uploaded.

**Comment by caraMsmith** [comment]
Ohhhh, yea I didn't refresh, I only reran the codes in R. Oh well
